research/global-config/index.md

```diff
@@ -0,0 +1,361 @@
+
+Flexible, config snippets can easily be removed using template conditional functionalities (as in the above example for the test job with trigger commit).
+
+##### Cons
+
+The templating syntax can be more error prone if compared with inheritance.
+
+#### Inheritance
+
+##### Pros
+
+Concise, it's the most concise syntax we could use and probably the least error prone.
+
+##### Cons
+
+Poor flexibility, I don't see an easy way to disable the above test job with trigger commit.
+
+Not really explicit, even though we use a placeholder it is harder to recognize the keys that need overriding.
+
+### Implementation
+
+Personally I find the _global config + overlays_ approach the best and in this case we would need to:
+
+- add the following keys to the `PackageConfig` class:
+
+```yml
+config:
+  base: https://gitlab.gnome.org/packit/templates/configs/default_packit.yaml.j2
+  values: ...
+```
+
+- load the packit.yaml file and search for the `config` key in it. If a `config` key is found we need to **recursively** look for the _base config_ and start processing all the templates in the chain, creating a new temporary packit.yml that will be used instead of the original one.
+  I see this code tied with the `LocalProject` class but I can be wrong.
+  We should make the new code work both for the packit CLI and the packit-service. Thinking at packit CLI, we should probably stay flexible and let the `base: URI` be also a local url (like `file:///`).
+
+- let the user know what the final configuration looks like (both for CLI and service).
+
+#### Jinja2 vs Ansible library
+
+For template management I would probably just use the **jinja2 template library**, even though we can also think about the ansible library.
+The ansible library could let us use `built-in filters and functions` but I don't see use cases for them and as a cons it has a heavier dependency footprint.
+
+### Performances
+
+Splitting the configuration in multiple configuration files will lead obviously to worst performance. Personally I don't see a way to prevent it.
+
+We can limit the number of recursion steps; 3/4 steps are, from my point of view, more than enough. Having a recursion limit will avoid an infinite recursion for malformed configurations.
+
+### packit-service defaults
+
+It could happen that the packit-service config defaults for the "Fedora CI instance" and those for the "Usual instance" (as an example) diverge.
+
+If it happens we could create two _hidden_, _inner_ packit config bases, one per instance, which will always be used for merging any packit config we process; in this way the differences will be grouped explicitly in a single place and we could, probably, enable and disable jobs for one instance or the other (as an example `pull-from-upstream` should not appear in fedora ci packit configuration) just using templating.
+
+<!--
+We already have a configuration file that enables packit we can still count on it.
```

Comment:
Since this allows multi-level approach, we can have something already for the Packit-owned configs. Similarly to your example above.

* packit-base (just to have something shared for all the packages if we need to influence anything. Maybe loaded by default for all the packages?)
* `packit-usecase` (`upstream-ci`, `downstream-ci`, `fedora-automation`), or `packit-instance`

Reply:
Looking at these Packit-owned ones. What do you think about supporting multiple base configs?

This might be handy to combine e.g. type of package (.eg. ecosystem-specific actions) with jobs setup (e.g.upstream ci).
If there is a list of base configs, we can just apply them one by one. But I get this might introduce too much mess into the system.